src/hazardous/base.rs

```diff
@@ -0,0 +1,156 @@
+use crate::errors::UnknownCryptoError;
+use std::{convert::TryFrom, fmt, marker::PhantomData};
+
+/// Marker trait for when a type contains some sensitive information.
+pub trait Secret {}
+
+/// Marker trait for when a type contains only non-sensitive information.
+/// Be careful if implementing this trait on your own. It cannot
+/// cause memory unsafety, and so is not marked `unsafe`. Implementing
+/// it can, however, lead to data types containing sensitive data ending
+/// up with APIs meant only for types containing only non-sensitive data.
+pub trait Public {}
+
+/// A small trait containing static information about the minimum and
+/// maximum size (in bytes) of a type containing data.
+pub trait Bounded {
+    /// The largest number of bytes this type should be allowed to hold.
+    const MIN: Option<usize> = None;
+
+    /// The smallest number of bytes this type should be allowed to hold.
+    const MAX: Option<usize> = None;
+}
+
+/// A generic holder for types that are basically just a bag of bytes
+/// with extra semantic meaning and restriction on top. We parameterize
+/// over the byte storage with parameter `B`. We parameterize over the
+/// API-level semantics of the type with phantom type `K`.
+#[derive(Clone)]
+pub struct Data<B, K> {
+    bytes: B,
+    phantom: PhantomData<K>,
+}
+
+impl<'a, B, K> Data<B, K>
+where
+    B: TryFrom<&'a [u8], Error = UnknownCryptoError>,
+    K: Bounded,
+{
+    /// TODO
+    pub fn from_slice(slice: &'a [u8]) -> Result<Self, B::Error> {
+        let min = K::MIN.unwrap_or(0);
+        let max = K::MAX.unwrap_or(usize::MAX);
+        if slice.len() < min || slice.len() > max {
+            return Err(UnknownCryptoError);
+        }
+
+        Ok(Self {
+            bytes: B::try_from(slice)?,
+            phantom: PhantomData,
+        })
+    }
+}
+
+impl<'a, B, K> Data<B, K>
+where
+    B: AsRef<[u8]>,
+{
+    /// Get the length of the contained byte slice.
+    pub fn len(&self) -> usize {
+        self.bytes.as_ref().len()
```

Comment:
We cannot get length based on the bytes here. If like BLAKE2b, it has valid ranges this will return the upper bound, because that's the array which has been allocated, not the actual length requested.

Reply:
I think I was assuming here that the B's AsRef implementation would take care of returning the correct subset. I was also assuming that B would be slightly more than an array. Something like (ignoring naming and syntax):

```rust
struct ArrayBytes {
    bytes: [u8: MAX],
    len: size,
}
```

Though I think it's definitely possible to just use basic types for B like `Data<[u8; 24]>` and then stick the extra logic in `Bound` maybe.

Reply:
I'm not sure at the moment, which one would work best tbh.

Reply:
I thought about this a bit more, and I see two separate use cases: an array of known size, and an array of known maximum size.

If we have a function that returns a precise number of bytes, we should probably use an `ArrayData` like:

```rust
struct ArrayData<const LEN: usize> {
   bytes: [u8; LEN]
}
```

and then define a new type like:

```rust
struct Sha256Ctx;
type Sha256Tag = PublicData<ArrayData<32>, Sha256Ctx>;
```

However, if we only know the *maximum* number of bytes to hold, maybe we define an `ArrayVecData`:

```rust
struct ArrayData<const MAX: usize> {
   bytes: [u8; MAX],
   len: usize,
}
```

and then define types like:

```rust
struct PasswordCtx;

// I don't know why we'd want to restrict the password length but let's pretend.
type Password = SecretData<ArrayVecData<32>, PasswordCtx>;

let pw0 = Password::from_slice(&b"rockyou");
```